### PR TITLE
feat(health-check): Smart QoS result on completed speed test

### PR DESCRIPTION
**Pull request by yarninisrael**
## Summary

Resolves the completed **Speed Test** page (Menu → Speed Test) into the **Smart QoS** feature surface, per Figma *Web-Light: Speed Test → Smart QoS* (node `21498:262334`):

- **Status card** with an Enable Smart QoS toggle + Active/Inactive pill
- **Protection Details** card — "Upload managed at X Mbps" / "No limit needed"
- **Advanced** disclosure — upload/download kbps, mode, overhead (read-only)
- Explainer card

Injected into `health_check/views/speed_test_view.dart` `_finishView` (desktop + mobile), shown only when a speed-test result exists.

## ⚠️ Needs a local compile before un-drafting

This was authored **without a local Flutter SDK**, and `main` has no PR-triggered analyze/test CI, so **nothing has compiled this yet**. Every widget API (`AppCard`, `AppSwitch`, `AppStatusLabel`, `AppExpansionCard`, `AppText`, `LinksysIcons`) was matched against real call sites on `main`, and `app_en.arb` validates as JSON — but please run `flutter pub get && flutter gen-l10n && flutter analyze && flutter test test/page/health_check/` before merge.

## ⚠️ Display + intent only — no backend

This firmware exposes **no QoS/shaping write path** (no CAKE/shaping JNAP action on `main`). So:
- The toggle **records intent only** — it does not write shaping config.
- Values come from `SmartQosRecommendation.fromResult`, a **client-side mirror** of the router's `smartqos.cgi` arithmetic (validated: 347337/7464 kbps → 6717/312603/UploadOnly).
- A user-facing note says shaping applies once the backend ships.
- **Delete the mirror model** (`smart_qos_recommendation.dart`) when a real transport lands and read values from the router instead.

## Files

- `lib/page/health_check/models/smart_qos_recommendation.dart` (+ unit tests)
- `lib/page/health_check/views/components/smart_qos_result.dart`
- `lib/page/health_check/views/speed_test_view.dart` — inject into `_finishView`
- `lib/l10n/app_en.arb` — 18 `smartQos*` keys + `mbps`

## Test plan

- [ ] `flutter analyze` clean on touched files
- [ ] `flutter test test/page/health_check/models/smart_qos_recommendation_test.dart`
- [ ] Visual check: Menu → Speed Test → run → confirm Smart QoS cards render (desktop + mobile)
- [ ] Non-English locale falls back to English for the 18 new keys (translations pending)

### CODE REVIEW

**Comment by yarninisrael** [comment]
# Auto QoS / Smart QoS — Handoff to Deven

**Goal for you:** produce **one complete PnP (plug-and-play) onboarding workflow that includes Auto QoS** in the Flutter firmware GUI (PrivacyGUI), building and shipping it through Linksys's own pipeline.

This document is the full story: what works, what doesn't, every issue we hit, what we tried, and exactly where things stand so you can pick it up without re-deriving it.

---

## TL;DR

- **Two separate things exist today.** (1) A working **standalone HTML** Auto QoS flow on the router (not Flutter). (2) A **Flutter** Smart QoS *result* page, ported to 1.3.0 and open as **draft PR #1363**.
- **What's NOT done — your task:** Auto QoS is **not** in the Flutter **PnP workflow**. Every attempt to put it there was blocked (details below).
- **The core blocker was never the Auto QoS code.** It was (a) self-built Flutter bundles breaking router login, and (b) our WIP being written against an architecture that doesn't exist on `main`.

---

## 1. What works right now

### 1a. Standalone HTML flow (on the router, shipping)
- URL: `https://192.168.1.1/smartqos/setup.html`
- Plain HTML/JS/CSS. Authenticates to a **CGI backend** (`/cgi-bin/smartqos.cgi`) via HTTP Basic auth — verified working (200 with correct pw, 401 without).
- Flow: GO dial → speed test (Cloudflare endpoints) → recommendation → enable → done.
- Repo: `github.com/yarninisrael/linksys-smart-qos` (files under `web/htdocs/`).
- **This is NOT part of PnP** and is NOT Flutter. It's a parallel deliverable that works because it sidesteps the Flutter login entirely.

### 1b. Flutter Smart QoS result page — **draft PR #1363**
- PR: https://github.com/linksys/PrivacyGUI/pull/1363 (base `main`, branch `feature/smart-qos-menu`)
- Adds the Smart QoS result design to the **Menu → Speed Test** page (`lib/page/health_check/views/speed_test_view.dart`, completed state).
- **Ported to the real 1.3.0 architecture** (`health_check` + JNAP `SpeedTestResult`), not the stack it was prototyped against.
- **Display + intent only** — see issue #4.
- **UNVERIFIED**: authored without a local Flutter SDK, and `main` has no PR-triggered analyze/test CI. Needs `flutter pub get && flutter gen-l10n && flutter analyze && flutter test test/page/health_check/` before merge.

---

## 2. The issues we hit (in the order they bit us)

### Issue #1 — Self-built Flutter bundles break router login (the big one)
- **Symptom:** after deploying a locally-built Flutter web bundle to the router's `/www`, login accepts the password but then bounces back to the login page with `#/localLoginPassword?error=noDeviceInfo`.
- **Root cause:** the password IS accepted, but the app then fails to establish its **USP device-info session** (`_fetchUspDeviceInfo()` throws), so it dumps back to login. This is a property of our **self-built** bundles, not the Auto QoS code and not the password.
- **Why it matters for you:** you cannot validate a PnP change by self-building and side-loading onto a router — you'll get locked out. **Build through Linksys CI** instead.
- **Recovery if it happens:** the router `/www` is overlayfs; pristine GUI is immutable in `/rom/www/`. Restore with `cp -a /rom/www/. /www/ && echo 3 > /proc/sys/vm/drop_caches`. This preserves `/www/smartqos/` and `/www/cgi-bin/smartqos.cgi`.

### Issue #2 — Our WIP was written against an architecture that doesn't exist on `main`
- The original WIP branch **`feature/smart-qos-jnap`** (commit `5d062fb8`) targeted `lib/page/unified_diagnostics/` (a `speed_test_notifier` / `SpeedTestState` / bps-based `SpeedTestResult` stack).
- **`origin/main` (Release 1.3.0) has no `unified_diagnostics/` directory at all.** On 1.3.0 the speed test lives in `lib/page/health_check/` on a different foundation (`HealthCheckState` with string statuses, JNAP `SpeedTestResult` in kbps, `privacygui_widgets`).
- Likewise the WIP's PnP code targeted `lib/page/instant_setup/providers/pnp_notifier.dart` — **that directory doesn't exist on `main` either.** 1.3.0 PnP uses `lib/page/instant_setup/data/pnp_provider.dart` (`BasePnpNotifier`) and a clean `PnpStep` abstraction (see §3).
- **Consequence:** the WIP is not cherry-pickable onto `main`. It's a reference for *intent and logic*, not code to merge. PR #1363 is the menu piece already re-ported; the PnP piece is still to do.

### Issue #3 — No local compile + no PR CI
- No Flutter SDK on the dev machine used, so nothing was compiled locally.
- `origin/main` has **no PR-triggered analyze/test workflow** (only: auto-add-to-project on issues, `deploy-demo` manual → public GH Pages needing `UI_KIT_DEPLOY_KEY`, generate-screenshots manual). So a PR to `main` does **not** auto-verify.
- **Consequence:** PR #1363 is a **draft** and needs a human with the SDK to compile it. Consider adding a lightweight PR CI (pub get + gen-l10n + analyze + test) as part of this work.

### Issue #4 — No QoS/shaping write path on the firmware GUI
- 1.3.0 has **no CAKE/shaping JNAP action** — the Flutter app cannot actually write shaping settings. (The only working shaping backend is the standalone page's `smartqos.cgi`, which the Flutter app doesn't speak to.)
- So in PR #1363 the toggle **records intent only**, and the recommendation values come from `SmartQosRecommendation.fromResult` — a **client-side mirror** of the router's `smartqos.cgi` arithmetic (validated: `347337/7464 kbps → 6717/312603/UploadOnly`).
- **For your PnP work:** decide the real transport. Either expose the shaping backend as a JNAP/USP action the Flutter app can call, or wire the app to the existing `smartqos.cgi`. Until then, "enable" can only record intent. **Delete the mirror model (`smart_qos_recommendation.dart`) once a real transport exists** and read values from the router.

### Issue #5 — Router serving gotchas (only relevant if you touch the standalone page)
- lighttpd on the router serves `.svg` as `application/octet-stream` (no MIME entry) → inline SVGs, don't `` them.
- The unit at `192.168.1.1` is not always the same physical device — fingerprint by hostname/uptime before deploying, and hash-verify every push.

---

## 3. Where to build the PnP integration on 1.3.0 (the actual target)

PnP on `main` uses a clean step abstraction — this is where Auto QoS should plug in:

- **Step contract:** `lib/page/instant_setup/model/pnp_step.dart` → `abstract class PnpStep` with `title()`, `onInit()`, `onNext()`, `save()`, `saveChanges`, `canGoNext()`.
- **Existing step impls:** `lib/page/instant_setup/model/impl/` — `personal_wifi_step.dart`, `guest_wifi_step.dart`, `night_mode_step.dart`, `your_network_step.dart`. **Copy one of these as the template for a `SmartQoSStep`.**
- **Step registration/order:** `lib/page/instant_setup/pnp_setup_view.dart` → `buildSteps()` returns `List<PnpStep>`. Insert `SmartQoSStep` as the **last step before final save** (so the speed test can run in the background during earlier steps and results are ready by the time the user reaches it).
- **State:** `lib/page/instant_setup/data/pnp_provider.dart` (`BasePnpNotifier`) + `pnp_state.dart` / `pnp_step_state.dart`.
- **Reusable UI:** the Smart QoS card widget from PR #1363 — `lib/page/health_check/views/components/smart_qos_result.dart` and the model `smart_qos_recommendation.dart`. Reuse these in the PnP step so the menu page and PnP show identical Smart QoS UI.

### Intended UX (from the original request)
- Start the speed test in the **background** when the user loads the first PnP page.
- On reaching the Smart QoS step: if the test is still running, show current stage; if it failed/stuck, offer configure-or-skip; if results are ready, present them with accept/reject.
- Smart QoS is the **last step before "saving your settings"** and before the home page.
- The final flow must terminate into the home page, not on the speed-test screen.

---

## 4. Design source (Figma)

- File: `i57ROb6ISj6wH3w7a8Zjzh` (Privacy GUI (Flutter) Web & Mobile)
- Speed-test-with-Smart-QoS page: node `1421-76471` → `Web-Light: Speed Test` frame; states:
  - `21498:262334` — Smart QoS (default)
  - `21498:262864` — Smart QoS - Advanced (disclosure expanded)
  - `21498:264003` — Smart QoS - Manual values applied
- PnP setup screens (intro/testing/toggle/enabled): nodes `21310:22994`, `21311:250153`, `21313:250429`, `21313:250730`.

---

## 5. Branches & PRs

| Ref | Where | What | Status |
|-----|-------|------|--------|
| PR **#1363** | `linksys/PrivacyGUI` | Smart QoS on **menu** speed-test page, ported to 1.3.0 | **Draft — needs compile** |
| `feature/smart-qos-menu` | `linksys/PrivacyGUI` | branch behind #1363 | pushed, in sync |
| `feature/smart-qos-jnap` (`5d062fb8`) | `linksys/PrivacyGUI` | **WIP reference** (menu + PnP) against the wrong arch | reference only, do NOT merge |
| `web/htdocs/*` | `yarninisrael/linksys-smart-qos` | standalone HTML flow (CGI-backed) | working on router |

---

## 6. Suggested plan for Deven

1. Compile-verify PR #1363 locally (`flutter analyze` + the health_check tests), fix anything, un-draft/merge. This gives a proven Smart QoS card widget to reuse.
2. Decide the shaping **transport** (issue #4) — JNAP/USP action or wire to `smartqos.cgi`. This unblocks a real "enable," not just intent.
3. Write `SmartQoSStep extends PnpStep` (copy an existing impl), reusing the PR #1363 widget/model.
4. Insert it into `buildSteps()` as the last step before save; start the speed test in the background at PnP entry.
5. Build/ship through Linksys CI — **do not self-build and side-load** (issue #1).

**Comment by yarninisrael** [comment]
@DevenDucommun — handing this off to you.

**Goal:** build **one complete PnP (plug-and-play) workflow with Auto QoS** in it.

This PR (#1363) is the **menu-page** Smart QoS result, ported to 1.3.0 and left as a **draft** because it hasn't been compiled (no local Flutter SDK here, and `main` has no PR analyze/test CI). Please run `flutter pub get && flutter gen-l10n && flutter analyze && flutter test test/page/health_check/` before un-drafting.

Full context — every issue we hit trying to get Auto QoS into the PnP workflow, what we tried, and exactly where things stand — is in the handoff report a couple comments up (also saved as `HANDOFF-DEVEN.md` in the standalone repo). Short version:

1. **Self-built Flutter bundles break router login** (`?error=noDeviceInfo`) → don't side-load; build via CI.
2. **Our WIP (`feature/smart-qos-jnap`, `5d062fb8`) targeted `unified_diagnostics/` + `instant_setup/providers/` — neither exists on `main`.** Reference only, not cherry-pickable.
3. **No QoS/shaping write path on the firmware GUI** → the toggle records intent only; values are a client-side mirror of `smartqos.cgi`. Needs a real transport.
4. **Where Auto QoS plugs into PnP on 1.3.0:** write a `SmartQoSStep extends PnpStep` (copy an impl from `lib/page/instant_setup/model/impl/`), register it in `pnp_setup_view.dart` → `buildSteps()` as the last step before save, and reuse the Smart QoS widget/model from this PR.

You've got push access to the standalone repo (`yarninisrael/linksys-smart-qos`) for the HTML flow. The PnP work lives here in `linksys/PrivacyGUI`. Ping if anything in the report is unclear.

File: lib/page/health_check/views/speed_test_view.dart [hunk 1]
```diff
@@ -585,6 +595,10 @@ class _SpeedTestViewState extends ConsumerState<SpeedTestView> {
                         uploadBandWidth.value,
                         uploadBandWidth.unit),
                     // ISP info
+                    if (smartQosRecommendation != null) ...[
+                      const AppGap.small3(),
+                      SmartQosResult(recommendation: smartQosRecommendation),
```
Comment (DevenDucommun) [inline lib/page/health_check/views/speed_test_view.dart:600]: **[blocking] Persist the user's choice or remove/disable this toggle for now.** Both desktop and mobile construct `SmartQosResult` without `initiallyEnabled` or `onEnabledChanged`, so the switch only changes the child widget's private state. On exact head `f44cc08c`, I exercised the UI by turning it on, unmounting the result, and revisiting it; the switch returned to off. There is therefore no recorded intent that can be “applied automatically” when a backend arrives, despite the user-facing note. Please either wire durable state/readback through these host call sites, or present this as a non-interactive preview until a real backend exists.

File: lib/page/health_check/models/smart_qos_recommendation.dart [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+  ///
+  /// The JNAP [SpeedTestResult] already reports bandwidth in kbps (the rest of
+  /// the health-check UI divides by 1000 to display Mbps), and latency in whole
+  /// milliseconds, so no unit conversion is needed before scaling.
+  factory SmartQosRecommendation.fromResult(SpeedTestResult result) {
+    final downloadKbpsMeasured = result.downloadBandwidth ?? 0;
+    final uploadKbpsMeasured = result.uploadBandwidth ?? 0;
+
+    return SmartQosRecommendation(
+      uploadKbps: uploadKbpsMeasured * shapingPercent ~/ 100,
```
Comment (DevenDucommun) [inline lib/page/health_check/models/smart_qos_recommendation.dart:80]: **[blocking] Do not generalize one observed router result into a user-facing recommendation contract.** The file itself lists the factor ordering, threshold basis, rounding, clamps, and WAN overhead as unvalidated, while the added tests mostly restate these constants and cover only one real fixture. The UI then presents the output as “Protection Details” / “Upload managed at” / “No limit needed”; there is no visible “estimate” qualifier even though the class comment says there is. Please validate the algorithm and boundary cases against the authoritative firmware implementation (with golden fixtures), or show only measured speed-test values until that contract exists.

File: lib/page/health_check/views/components/smart_qos_result.dart [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+                    AppText.bodySmall(
+                      loc(context).smartQosEnableDescription,
+                      color: Theme.of(context).colorScheme.onSurface,
+                    ),
+                  ],
+                ),
+              ),
+              const AppGap.medium(),
+              AppSwitch(
+                semanticLabel: 'enable smart qos',
```
Comment (DevenDucommun) [inline lib/page/health_check/views/components/smart_qos_result.dart:105]: **[non-blocking] Localize this accessibility label.** The visible control text uses `loc(context)`, but its semantic label is permanently English. Reuse the localized enable label (or add a dedicated ARB key) so screen-reader output follows the selected locale.

**Review by DevenDucommun — COMMENTED** [review]
Reviewed exact head `f44cc08c5da14d0baf589cb90a8e899e212b68d3` with the repository-pinned Flutter 3.27.2 SDK.

Two blocking correctness/product findings are inline: the Enable control does not record intent, and the UI promotes an algorithm with explicitly unvalidated boundaries into definitive recommendations. I also left a non-blocking localization/accessibility comment.

- Security: no new credential, transport, or router-write path is introduced; keeping this display-only is a good boundary.
- Correctness: the added model tests pass, but they validate the client constants rather than the unknown firmware contract. A disposable widget test confirmed enable state is lost after the result widget is revisited.
- Performance: the added arithmetic and widgets are small; no material performance concern found.
- Maintainability: the temporary mirror is clearly documented, but its undetectable firmware-drift risk is real; the touched-file analyzer also reports the new `AppSwitch` import as unnecessary.

Verification performed: `flutter pub get`, `flutter gen-l10n`, targeted `flutter test test/page/health_check/models/smart_qos_recommendation_test.dart` (8/8 pass), a disposable enable/revisit widget probe (pass, demonstrating the reset), and targeted `flutter analyze` (the new widget compiles; one new unnecessary-import info plus pre-existing warnings in `speed_test_view.dart`). The PR correctly remains draft.
